### ISSUE TITLE
Create SCons syntax/extension for Visual Studio Code

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#3210]
Create a plugin for Visual Studio code which does syntax completion/highlighting

As an example of how to do such:
https://github.com/twxs/vs.language.cmake

**Comment by mwichmann** [comment]
Sure, but then the question becomes... what about Atom? What about PyCharm? What about SublimeText? what about ... ?  fwiw, looks like atom has a tiny bit of support... not syntax highlighting, but at least the ability to execute an scons-managed project, that's a start... https://atom.io/packages/build-scons.

**Comment by bdbaddog** [comment]
Looks like it's possible to provide support for all with single file. Textmate compatible syntax file seems to work with almost every thing.. More details need to be researched though.

**Comment by creikey** [comment]
Couldn't you just treat the file as a python file and have the python extension provide the support?

**Comment by bdbaddog** [comment]
The issue would be autocompletion and highlighting scons keywords.

**Comment by mwichmann** [comment]
Yes, easy enough to teach it (as I have done in copy of PyCharm) that files names SConstruct and SConscript (and for our case *.scons) are Python, then you get python syntax stuff. But in any "smart" environment like this, all scons API ends up being flagged as unknown, undefined references, etc. which makes it a lot less useful.

**Comment by bdbaddog** [comment]
The syntax highlighting and autocompletion is really the useful bit. Teaching any editor SConscript and SConstruct are python is not too difficult. (I've done the same in my pycharm setup)

**Comment by aaronfranke** [comment]
>what about Atom? What about PyCharm? What about SublimeText? what about ... ?

Supporting any is better than supporting none. VS Code is quite popular, so it's a good first choice.

**Comment by philosophicles** [comment]
For reference, any VS Code user can [follow these 10-second instructions](https://code.visualstudio.com/docs/languages/overview#_adding-a-file-extension-to-a-language) to tell their installation that `SConstruct` and `SConscript` files contain python. 

I have raised an issue to the VS Code team to add that recognition by default, as I think it'd be helpful and non-destructive: https://github.com/microsoft/vscode/issues/105623. Please feel free to upvote if you agree.

As @mwichmann said above, if you have a python linter enabled, doing this will start flagging errors in the `SConstruct` and `SConscript` files, because Scons API objects aren't explicitly imported. For simple SCons configuration you could disable the linter, either for the whole VSCode project (via F1 menu) or per-file ([e.g. for PyLint](https://stackoverflow.com/questions/10238473/pylint-disable-all-warnings-for-a-file)). With per-file exclusions you could still lint utility scripts etc. 

I think a SCons extension would then only really be useful if you have a complex SCons scenario where the main `SConstruct` and `SConscript` files need linting/validating.

**Comment by stefanobieler** [comment]
This would definitely enhance my quality of life. please stop this pain! xD

**Comment by ghost** [comment]
Can you import objects to cover these keywords?

**Comment by bdbaddog** [comment]
> Can you import objects to cover these keywords?

It's not that simple. The Globals passed to eval() when SCons processes SConstruct/SConscripts are explicitly created. So you'd need to probably dump from a fully populated global to create some type of syntactical stub file for the editors to use I think.

**Comment by mwichmann** [comment]
might be worth thinking about that anyway... some people now advocate having a specific thing you import that gives you "the API".  I'm not sure if the current recommendation (`from SCons.Script import *`) is precise enough about that...

**Comment by bdbaddog** [comment]
> might be worth thinking about that anyway... some people now advocate having a specific thing you import that gives you "the API". I'm not sure if the current recommendation (`from SCons.Script import *`) is precise enough about that...

I guess there's two possible levels of functionality:

1. enough support that editors can auto complete
2. enough support for above plus being able to follow code through to definitions..

**Comment by ghost** [comment]
I personally would like just enough so that vscode doesn't mark them as errors, I don't need anything else

**Comment by bdbaddog** [comment]
> I personally would like just enough so that vscode doesn't mark them as errors, I don't need anything else

Indeed.
Contributions welcome.  If you have any idea how to create vs code (or other IDE) plugin for such, we could use the help.

**Comment by mwichmann** [comment]
There's a slack channel that's reputed to be the best place to get help... I could never make any headway figuring out what was needed, all the examples of extending Code didn't seem to apply to our case.

https://join.slack.com/t/vscode-dev-community/shared_invite/zt-ggynfxra-KAozfLTWe03w_PwRxwtfvA

**Comment by SirUppyPancakes** [comment]
One way to get vscode to recognize the extension-less `SConstruct` files as python is to use a shebang at the top (`#!/usr/bin/env python`), and vscode will automatically detect it.

I'm rather new and just digging into SCons now, but all of the SCons-specific stuff on top of Python is just a bunch of "magic imports" right? Like when you run a SConscript file with `scons` instead of `python`, it's as if you've just imported some symbols from some modules by default. Given that, I did a quick test where I copied the `SCons` folder from the SCons source to my root project directory (same level as the `SConstruct` file), and then added the following to the top of my SConstruct file:
```python
import sys

sys.path.append("./SCons")

from SCons.Script import *
```

This causes all of the built-in symbols from SCons to get picked up and you can hover over them and navigate to their source just like any normal python package. What I haven't done is try and run with this change, and I'm not sure if importing like this would break something. But this makes me think that we could probably distribute the source with the SCons package and allow users to import explicitly if they want the full syntax highlighting? Or I suppose we could use stub files, though I'm not super familiar with how that works to be honest. The benefit to this approach of course is that it would work with any IDEs that support python automatically.

**Comment by bdbaddog** [comment]
> One way to get vscode to recognize the extension-less `SConstruct` files as python is to use a shebang at the top (`#!/usr/bin/env python`), and vscode will automatically detect it.
> 
> I'm rather new and just digging into SCons now, but all of the SCons-specific stuff on top of Python is just a bunch of "magic imports" right? Like when you run a SConscript file with `scons` instead of `python`, it's as if you've just imported some symbols from some modules by default. Given that, I did a quick test where I copied the `SCons` folder from the SCons source to my root project directory (same level as the `SConstruct` file), and then added the following to the top of my SConstruct file:
> 
> ```python
> import sys
> 
> sys.path.append("./SCons")
> 
> from SCons.Script import *
> ```
> 
> This causes all of the built-in symbols from SCons to get picked up and you can hover over them and navigate to their source just like any normal python package. What I haven't done is try and run with this change, and I'm not sure if importing like this would break something. But this makes me think that we could probably distribute the source with the SCons package and allow users to import explicitly if they want the full syntax highlighting? Or I suppose we could use stub files, though I'm not super familiar with how that works to be honest. The benefit to this approach of course is that it would work with any IDEs that support python automatically.

The import would get you some of the way there, but there's more. 
Also no need to add to sys.path, some IDE's should pick up scons from the python you've set them up to use.

**Comment by SirUppyPancakes** [comment]
Ah yeah oops looks like my initial test misled me to believe I needed the source, and `from SCons.Script import *` does the trick.

Can you point me to an example of the "more" you're referring to? From a brief looking a the source, seems like there are some usages of `exec` and other dynamic features which would make relying on simple Python autocomplete difficult? I'd also be curious to know about the reasoning behind going that route, rather than choosing to use more basic Python features which would allow us to more fully take advantage of the language and support it has across IDEs.

Also, let me know if I'm getting too off topic for this thread, and I'd be happy to go ask in the Slack channel or other recommended channel.

**Comment by mwichmann** [comment]
Discord channel... find link at https://scons.org/contact.html

**Comment by mwichmann** [comment]
It's not really off-topic to mention why *good* IDE support will be tricky... 
* The most important thing people do in their build setup is call Builders - the things that establish the relationship between sources and targets.  All Builders are dynamically created, through a runtime call to the `Builder` factory function. Since there's no static definition, that's hard for an IDE to work with.  I'm wondering if manually creating a typing stub for the builders, "as they would be once created" could help.  Why is this so? That's for another discussion :)
* SCons in general is quite liberal in accepting undefined keyword arguments in various calls (often factory functions), and just letting something sensible be done with them.  The most notable example is `Environment`, where keyword arguments not specifically recognizes just get added as entries in the environment thus created. That makes it impossible for an IDE to tell you that you have misspelled a kwarg - for example, it's quite easy to type `CPPPATHS` when you meant `CPPPATH` - `CPPPATHS` will just be entered as a new construction variable which nothing will ever use. 
* SCons code way precedes Python's introduction of keyword-only and positional-only arguments, and so accepts both forms in many places where it would be nicer to take just one - this would reduce an IDE's ability to give useful hints.
* SCons does a lot of... well, let's call it "dodgy" fiddling with things, taking more advantage of Python's dynamic nature than one might want. Types morph themselves into other types, tons of things get created in class instances outside of the initializer, and lots more. I know an IDE would have trouble with those because static type checkers keep yelling about these things, although they're perfectly "legal" in Python.  This flexible usage also makes it very hard to add type annotations in many places (something we're doing on a very gradual basis) - parameters can often be of a multitude of different types, some of them non-obvious.

There's probably more...
